### DISCUSSION TITLE
Improving the performance of `jobrunr_jobs_stats` view

### DISCUSSION THREAD

**Discussion by MikeyBower93** [jobrunr/jobrunr#834] [Ideas]
Hey team!

I've been inspecting some of the JobRunr queries on our database to analyse where performance could be improved and noticed this query
```sql
select * from jobrunr_jobs_stats ORDER BY total ASC LIMIT ? OFFSET ?
```
Looking into this, this query is executing frequently because we added a `JobStatsChangeListener` to our code so we could report job successes/failures to Datadog and create graphs to explain our performance. However I can see the query for this is on average 4 times slower than the rest of JobRunr's database queries. On our database its performing with a P50 latency of 91ms, and P99 latency of 450ms, which comparative to other queries seems quite slow.

Looking into this, it seems `jobrunr_jobs_stats` is a view that runs aggregations on the `jobrunr_jobs` table, and whilst it does use indexes (index only scans, to get this data, my understanding is it still has to some quite heavy performing to get the counts as it needs to scan all the leaf nodes in the index, which for deleted jobs is often a lot and is not very selective. 

Here is the execution plan
```json
[{"Plan":{"Node Type":"Aggregate","Strategy":"Sorted","Partial Mode":"Simple","Parallel Aware":false,"Startup Cost":0.42,"Total Cost":82780.57,"Plan Rows":4,"Plan Width":16,"Actual Startup Time":127.754,"Actual Total Time":188.741,"Actual Rows":4,"Actual Loops":1,"Grouping Sets":[{"Group Keys":[["state"],[]]}],"Shared Hit Blocks":42544,"Shared Read Blocks":0,"Shared Dirtied Blocks":0,"Shared Written Blocks":0,"Local Hit Blocks":0,"Local Read Blocks":0,"Local Dirtied Blocks":0,"Local Written Blocks":0,"Temp Read Blocks":0,"Temp Written Blocks":0,"I/O Read Time":0,"I/O Write Time":0,"Plans":[{"Node Type":"Index Only Scan","Parent Relationship":"Outer","Parallel Aware":false,"Scan Direction":"Forward","Index Name":"jobrunr_state_idx","Relation Name":"jobrunr_jobs","Alias":"jobrunr_jobs","Startup Cost":0.42,"Total Cost":80992.35,"Plan Rows":357636,"Plan Width":8,"Actual Startup Time":0.03,"Actual Total Time":136.606,"Actual Rows":353287,"Actual Loops":1,"Heap Fetches":113940,"Shared Hit Blocks":42544,"Shared Read Blocks":0,"Shared Dirtied Blocks":0,"Shared Written Blocks":0,"Local Hit Blocks":0,"Local Read Blocks":0,"Local Dirtied Blocks":0,"Local Written Blocks":0,"Temp Read Blocks":0,"Temp Written Blocks":0,"I/O Read Time":0,"I/O Write Time":0}]},"Planning":{"Shared Hit Blocks":131,"Shared Read Blocks":0,"Shared Dirtied Blocks":0,"Shared Written Blocks":0,"Local Hit Blocks":0,"Local Read Blocks":0,"Local Dirtied Blocks":0,"Local Written Blocks":0,"Temp Read Blocks":0,"Temp Written Blocks":0,"I/O Read Time":0,"I/O Write Time":0},"Planning Time":0.343,"Triggers":[],"Execution Time":188.777}]
```

We can see that its doing a index only scan, but still has a cost of 81,000 which is seemingly quite high, due to nature of a count on a non selective index condition.

I wondered if some considerations could be made to improving this? Perhaps you could maintain a table that stores these raw numbers that isn't a view, and use some triggers to increment/decrement when particular events happen?

For context, this runs a service that has processed 40 million jobs and counting.

Thanks!
Mike

**Comment by rdehuyss** [comment]
Hi @MikeyBower93 , yes - we're aware of this and already tried to tune this query as much as possible (indices, ... ).

Two options that I can give currently:
- only enable your JobStatsChangeListener on the server that is running your dashboard. This way, the query is only done once as it is also used by the dashboard. 
- limit the amount of deleted and succeeded jobs (in JobRunr Pro you can configure the moment a Job is deleted __*per job*__). In JobRunr OSS this can be configured globally.

An item exists on our JobRunr Pro backlog to improve this using counters (it has not yet been done as I'm a bit scared of having data that is out-of-sync). This improvement is planned for early next year.

**Reply by MikeyBower93** [reply]
Awesome! I will take one of those initial steps for now, but thanks for confirming its on your roadmap 😊

**Comment by ilinic** [comment]
Ronald, is it safe to reduce rateLimit of this query from 1 request per second to 1 request per 10 seconds?

**Reply by rdehuyss** [reply]
Hi @ilinic , yes, that's definitely an option.

As an FYI:
- we've improved the performance of this query in JobRunr v7. 
- we also tried to implement this with counters using stored procedures but as soon as we added these stored procedures, job throughput was only 25% compared to a database without stored procedures. So this is not an option for us.
- we still have other idea's we would like to try but did not make it in v7

**Comment by mhdatie** [comment]
Hey Ronald, we have been experiencing very slow execution times from this query on very large Jobrunr job tables using Jobrunr Pro with Quarkus `org.jobrunr:quarkus-jobrunr-pro:7.5.2`. 

Here's the state of the deployment
- We have the below metrics config enabled on all of our workers despite having only a number of them configured with an enabled background server during runtime (we use a multi-purpose container image so Jobrunr is running alongside other components)
   - ~I believe these configs are build time configs and not runtime so we have to set them during compilation, but I can confirm~
- Only two workers have the dashboard enabled during runtime

```
quarkus.jobrunr.background-job-server.metrics.enabled=true
quarkus.jobrunr.jobs.metrics.enabled=true
```

Query executed:
```sql
SELECT
  *
FROM
  jobrunr_jobs_stats
```

For example, with a long running jobs which eventually process 2M jobs, and with a Postgres instance with 8 vCPUs and 50% memory utilization, this query takes up 100% CPUs

<img width="766" height="727" alt="Screenshot 2025-07-16 at 11 16 12 AM" src="https://github.com/user-attachments/assets/758da0f0-2bda-4b77-a5ef-b09aac06905d" />

<img width="1199" height="244" alt="Screenshot 2025-07-16 at 11 11 58 AM" src="https://github.com/user-attachments/assets/49b051d5-3c19-4f44-a210-e144d97c8775" />

Any ideas how we can help improve this query or the deployment configuration? Any noticeable improvements with this query with the v8 upgrade?

Thanks!

**Comment by mhdatie** [comment]
Update - We were able to reduce the frequency at which this query runs by only setting `quarkus.jobrunr.jobs.metrics.enabled` to true in the Jobrunr dashboard worker. This means that other workers which save jobs to the jobrunr table won't be able to invoke the `getJobStats` call since there would be no metric binder listener on them. 

I think this piece of information is lacking in the docs and I was only able to find it by digging through the source code. 

However, we still experience high CPU usage whenever this job stats query runs and a lot of jobs are constantly being enqueued or while jobs which involve a CPU intensive query are running.

**Reply by mhdatie** [reply]
Another option which could be a quick big win, is to make the timer at which these stat queries run configurable rather than statically set to 5 seconds in the AbstractStorageProvider class.

**Comment by rdehuyss** [comment]
Hey @mhdatie , I'm ooo for the moment and not checking GitHub.

Will get back to you somewhere in August.

**Comment by rdehuyss** [comment]
Pinging @wgroeneveld - can you help here?

@mhdatie : this is probably miss configuration. The metrics should only be enabled on the dashboard server. 
The rate limiter allows to configure how many times the query is executed.

**Reply by mhdatie** [reply]
Thanks Ronald, we have fixed our configuration to only apply `quarkus.jobrunr.jobs.metrics.enabled` in the dashboard. That dropped the number of times we make the query request significantly.

`quarkus.jobrunr.background-job-server.metrics.enabled` was only working if it was enabled alongside the background job server -
https://github.com/jobrunr/jobrunr/blob/v7.5.2/framework-support/jobrunr-quarkus-extension/quarkus-jobrunr/src/main/java/org/jobrunr/quarkus/autoconfigure/metrics/JobRunrMetricsProducer.java#L49


> The rate limiter allows to configure how many times the query is executed.

Which rate limiter are you referring to? Any docs around that?

**Reply by AlessandroBagnoli** [reply]
Hi @rdehuyss, how one can configure the rate limiter for limiting how many times the query is executed? This is quite expensive in one of our services. I think it's because we have many (+100k) scheduled jobs, hence the query is slow in counting those scheduled jobs. 

If the query is ran just for dashboard purposes, I am more than fine in limiting its rate of execution to, say, once every 5s or even 10s. Is that possible?

**Reply by rdehuyss** [reply]
Hi @AlessandroBagnoli, @mhdatie - currently, the Job Stats view should only be queried once every 5 seconds for the dashboard so we don't understand yet why this is causing so much load.

Are you sure you're not exposing the metrics to MicroMeter / Prometheus somehow?

**Reply by AlessandroBagnoli** [reply]
> Hi @AlessandroBagnoli, @mhdatie - currently, the Job Stats view should only be queried once every 5 seconds for the dashboard so we don't understand yet why this is causing so much load.
> 
> Are you sure you're not exposing the metrics to MicroMeter / Prometheus somehow?

Yes, actually:

```
        .useMicroMeter(
            JobRunrMicroMeterIntegration(
                prometheusMeterRegistry,
                // publish job metrics
                true,
            ),
        )
```

@rdehuyss Do you think setting the flag to `false` should lower the frequency of that query, at the cost of a little less observability?

**Reply by rdehuyss** [reply]
Yes, this is the culprit. If all your `BackgroundJobServer` instances are doing this, they are all doing the same heavy query but they are all reporting the exact same data. 

The `JobRunrMicroMeterIntegration` may only be configured/integrated on the server that is running the dashboard. That should solve your issue.

**Reply by AlessandroBagnoli** [reply]
We removed the micrometer integration and already seeing the benefits:

<img width="2544" height="978" alt="image" src="https://github.com/user-attachments/assets/7eef58ea-e6ea-4873-bd60-5dee387999f6" />

Thanks a lot.
